Systemctl show json output

phomes opened the pull request:
I was going through the TODO looking for something easy to implement.

This extracts the bus message to json transform code from busctl, makes the type variant string optional, and makes use of the transform in "systemctl show".

Is this generally going in the right direction?

The non-json output in print_property() prints values for a few properties in a human friendly way. I was not sure if this was desirable for the json output as well, or if it should stay more true to the values in the bus message. I went with the latter.

phomes commented:
> This pull request **introduces 1 alert** when merging [1130145](https://github.com/systemd/systemd/commit/113014517e39e85659897de6656fa3aea3633c9e) into [8e45c72](https://github.com/systemd/systemd/commit/8e45c72cf581a2224725469376d13ca4dcd77a74) - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-956ae7e6008acec70161b1db50c621bca59cdecf)
> 
> **new alerts:**
> 
> * 1 for Comparison result is always the same

Force pushed a fix for this

bluca commented on src/shared/json-transform.h:
```diff
@@ -0,0 +1,4 @@
+#include "bus-message.h"
+#include "json.h"
+
+int json_transform_message(sd_bus_message *m, JsonVariant **ret);
```
nit: missing newline at the end of the file

phomes replied:
fixed. This revealed that I could not compile after this commit. The neccesary change was included in the following commit. I also corrected that.

bluca commented on src/systemctl/systemctl.h:
```diff
@@ -92,3 +92,4 @@ extern char **arg_clean_what;
 extern TimestampStyle arg_timestamp_style;
 extern bool arg_read_only;
 extern bool arg_mkdir;
+extern JsonFormatFlags arg_json_format_flags;
```
newline missing again

phomes replied:
fixed

bluca commented:
Nice! So it prints all properties of a unit, the --property switch does not work - is that intended?

phomes commented:
> Nice! So it prints all properties of a unit, the --property switch does not work - is that intended?

To be honest it was just an oversight. Asking for a single property formatted as json seems a bit weird though, and perhaps not worth the added complexity in the transform code?

bluca commented:
> > Nice! So it prints all properties of a unit, the --property switch does not work - is that intended?
> 
> To be honest it was just an oversight. Asking for a single property formatted as json seems a bit weird though, and perhaps not worth the added complexity in the transform code?

That's fine, but then --property should be refused together with --json for the 'show' verb, I think

boucman commented:
using --json with multiple --property as a way to filter would make sense, though...

bluca commented on TODO:
```diff
@@ -523,10 +523,6 @@ Features:
 * sd-boot: optionally, show boot menu when previous default boot item has
   non-zero "tries done" count
 
-* introduce an option (or replacement) for "systemctl show" that outputs all
```
Can you leave a to-do to support --properties as well?

keszybz commented on src/shared/json-transform.h:
```diff
@@ -0,0 +1,5 @@
+#include "bus-message.h"
```
SPDX...

keszybz commented on src/shared/json-transform.h:
```diff
@@ -0,0 +1,5 @@
+#include "bus-message.h"
+#include "json.h"
+
```
`#pragma once`

keszybz commented on src/shared/json-transform.h:
```diff
@@ -0,0 +1,5 @@
+#include "bus-message.h"
+#include "json.h"
+
+int json_transform_message(sd_bus_message *m, JsonVariant **ret);
```
This also needs `#incldue <sd-bus.h>`

keszybz commented on src/shared/json-transform.h:
```diff
@@ -1,5 +1,11 @@
-int json_transform_variant(sd_bus_message *m, const char *contents, JsonVariant **ret);
+typedef enum JsonTransformFlags {
+        JSON_TRANSFORM_PLAIN = 1 << 0,
+        JSON_TRANSFORM_TYPE_DATA  = 1 << 1, /* Insert layer with variant type/data */
```
I'd prefer to have the equal signs aligned. Looks nicer when there's more entries.

keszybz commented on src/shared/json-transform.h:
```diff
@@ -1,5 +1,11 @@
+typedef enum JsonTransformFlags {
+        JSON_TRANSFORM_PLAIN = 1 << 0,
+        JSON_TRANSFORM_TYPE_DATA  = 1 << 1, /* Insert layer with variant type/data */
+
```
Unnecessary empty line here.

keszybz commented on src/systemctl/systemctl-show.c:
```diff
@@ -1889,13 +1891,32 @@ static int show_one(
-        r = bus_message_print_all_properties(reply, print_property, arg_properties, arg_value, arg_all, &found_properties);
-        if (r < 0)
-                return bus_log_parse_error(r);
+        if (!FLAGS_SET(arg_json_format_flags, JSON_FORMAT_OFF)) {
```
Why not revert the codnition (and the order of branches)? Positive conditions are easier to think about...

keszybz commented on src/systemctl/systemctl-show.c:
```diff
@@ -1889,13 +1891,32 @@ static int show_one(
+
+                if (arg_properties)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                "--property is not supported when outputting json");
```
This check should be done when parsing args, not here, after some real work has been done.

phomes replied:
implemented --property so no longer needed

phomes commented:
Thanks for the reviews. I fixed the comments and implemented support for --property.

Last thing to consider is whether the json output should be based on the content of the bus message, or if we should apply the same pretty printing as we do in "systemctl show" today. Most properties can be printed without any problem, but e.g. ExecStart has a lot of pretty print:

```
systemctl show acpid --property ExecStart
ExecStart={ path=/usr/sbin/acpid ; argv[]=/usr/sbin/acpid $OPTIONS ; ignore_errors=no ; start_time=[Fri 2021-02-12 21:10:42 CET] ; stop_time=[n/a] ; pid=823 ; code=(null) ; status=0/0 }
```

```
systemctl show acpid --property ExecStart --json=pretty
{
        "ExecStart" : [
                [
                        "/usr/sbin/acpid",
                        [
                                "/usr/sbin/acpid",
                                "$OPTIONS"
                        ],
                        false,
                        1613160642256769,
                        6690114,
                        0,
                        0,
                        823,
                        0,
                        0
                ]
        ]
}

```

bluca commented:
Probably best to leave the raw data? json output is for machine consumption usually

bluca commented:
@phomes there are merge conflicts, could you please push a rebase?

poettering reviewed (changes requested):
looks pretty good!

poettering commented:
So, in light of #19451 I wonder if this still is the way to go. I think it would be much nicer to have *native* JSON data exported by PID 1, via a Describe() bus call on the Unit object, and then have systemctl just output that. i.e. move json generation server-side, and use D-Bus as a simple transport for it.

YHNdnzj commented:
As discussed in https://github.com/systemd/systemd/pull/29648 and #31948, we shall instead utilize varlink to generate native json outputs in pid1. Closing this one hence.
